06_gpu_and_ml/embeddings/wikipedia/main.py

```diff
@@ -0,0 +1,327 @@
+    # "ghcr.io/huggingface/text-embeddings-inference:0.3.0"  # Turing for T4s.
+)
+
+## Dataset-Specific Configuration
+DATASET_NAME = "wikipedia"
+DATASET_READ_VOLUME = Volume.persisted("embedding-wikipedia")
+EMBEDDING_CHECKPOINT_VOLUME = Volume.persisted("checkpoint")
+DATASET_DIR = "/data"
+CHECKPOINT_DIR = "/checkpoint"
+SAVE_TO_DISK = True
+
+## Upload-Specific Configuration
+DATASET_HF_UPLOAD_REPO_NAME = "567-labs/upload-test"
+UPLOAD_TO_HF = True
+
+## HF Text-Embedding Inference specific Configuration
+
+LAUNCH_FLAGS = [
+    "--model-id",
+    MODEL_ID,
+    "--port",
+    "8000",
+    "--max-client-batch-size",
+    str(BATCH_SIZE),
+    "--max-batch-tokens",
+    str(BATCH_SIZE * 512),
+]
+
+
+stub = Stub("embeddings")
+
+
+def spawn_server() -> subprocess.Popen:
+    import socket
+
+    process = subprocess.Popen(["text-embeddings-router"] + LAUNCH_FLAGS)
+    # Poll until webserver at 127.0.0.1:8000 accepts connections before running inputs.
+    while True:
+        try:
+            socket.create_connection(("127.0.0.1", 8000), timeout=1).close()
+            print("Webserver ready!")
+            return process
+        except (socket.timeout, ConnectionRefusedError):
+            # Check if launcher webserving process has exited.
+            # If so, a connection can never be made.
+            retcode = process.poll()
+            if retcode is not None:
+                raise RuntimeError(f"launcher exited unexpectedly with code {retcode}")
+
+
+def download_model():
+    spawn_server()
+
+
+tei_image = (
+    Image.from_registry(
+        "ghcr.io/huggingface/text-embeddings-inference:86-0.4.0",
+        add_python="3.10",
+    )
+    .dockerfile_commands("ENTRYPOINT []")
+    .run_function(download_model, gpu=GPU_CONFIG)
+    .pip_install("httpx")
+)
+
+with tei_image.imports():
+    import numpy as np
+
+
+def generate_chunks_from_dataset(xs, chunk_size: int):
+    """
+    Generate chunks from a dataset.
+
+    Args:
+        xs (list): The dataset containing dictionaries with "id", "url", "title", and "text" keys.
+        chunk_size (int): The size of each chunk.
+
+    Yields:
+        tuple: A tuple containing the id, url, title, and a chunk of text.
+
+    """
+    for data in xs:
+        id_ = data["id"]
+        url = data["url"]
+        title = data["title"]
+        text = data["text"]
+        for chunk_start in range(0, len(text), chunk_size):
+            yield (
+                id_,
+                url,
+                title,
+                text[chunk_start : chunk_start + chunk_size],
+            )
+
+
+def generate_batches(xs, batch_size):
+    batch = []
+    for x in xs:
+        batch.append(x)
+        if len(batch) == batch_size:
+            yield batch
+            batch = []
+    if batch:
+        yield batch
+
+
+@stub.cls(
+    gpu=GPU_CONFIG,
+    image=tei_image,
+    concurrency_limit=N_GPU,
+    retries=3,
+)
+class TextEmbeddingsInference:
+    def __enter__(self):
```

Comment:
It could be quite nice to add a comment here that the Modal class syntax allows you to define a complex container lifecycle in one fell swoop.

The `__enter__` function runs once per GPU startup, and the `@method()` function runs once per inference on an initialized GPU.

Reply:
Also, is there a reason we left out `allow_concurrent_inputs` here?